File: plugins/serialization/files/utterance/ebml/swig/python/utt_ebml.py
######################################################################################
## Copyright (c) 2009 The Department of Arts and Culture,                           ##
## The Government of the Republic of South Africa.                                  ##
##                                                                                  ##
## Contributors:  Meraka Institute, CSIR, South Africa.                             ##
##                                                                                  ##
## Permission is hereby granted, free of charge, to any person obtaining a copy     ##
## of this software and associated documentation files (the "Software"), to deal    ##
## in the Software without restriction, including without limitation the rights     ##
## to use, copy, modify, merge, publish, distribute, sublicense, and#or sell        ##
## copies of the Software, and to permit persons to whom the Software is            ##
## furnished to do so, subject to the following conditions:                         ##
## The above copyright notice and this permission notice shall be included in       ##
## all copies or substantial portions of the Software.                              ##
##                                                                                  ##
## THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR       ##
## IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,         ##
## FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE      ##
## AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER           ##
## LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,    ##
## OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN        ##
## THE SOFTWARE.                                                                    ##
##                                                                                  ##
######################################################################################
##                                                                                  ##
## AUTHOR  : Aby Louw                                                               ##
## DATE    : December 2009                                                          ##
##                                                                                  ##
######################################################################################
##                                                                                  ##
## Python functions for loading/saving utterances in EBML format.                   ##
##                                                                                  ##
##                                                                                  ##
######################################################################################



%pythoncode
%{
def load_ebml_utterance(path):
    """
    Load an EBML format utterance from the given file name.
    @param path: The full path and filename of the utterance to load.
    @type path: str
    @return: The loaded utterance.
    @rtype: L{SUtterance}
    """

    if not isinstance(path, str):
        raise TypeError('path must be of type \'str\'')

    utt_object = py_sutterance_ebml_load(path)
    if not utt_object:
        string = "Speect failed to load utterance of type \'EBML\' from file \'" +\
                 path + "\'"
        raise RuntimeError(string)

    return speect.SUtterance(object=utt_object, owner=True)


def save_ebml_utterance(self, path):
    """
    Save an EBML format utterance to the given file name.
    @param path: The full path and filename where to save the utterance.
    @type path: str
    """

    if not isinstance(path, str):
        raise TypeError('path must be of type \'str\'')

    utt_object = self._get_speect_object()
    if not utt_object:
        raise RuntimeError("Utterance has no Speect object")

    py_sutterance_ebml_save(utt_object, path)
    
# add the functions to the Speect SUtterance class
setattr(speect.SUtterance, "load_ebml", staticmethod(load_ebml_utterance))
setattr(speect.SUtterance, "save_ebml", save_ebml_utterance)
%}

File: plugins/serialization/files/utterance/ebml/swig/python/utt_ebml.c
/************************************************************************************/
/* Copyright (c) 2009 The Department of Arts and Culture,                           */
/* The Government of the Republic of South Africa.                                  */
/*                                                                                  */
/* Contributors:  Meraka Institute, CSIR, South Africa.                             */
/*                                                                                  */
/* Permission is hereby granted, free of charge, to any person obtaining a copy     */
/* of this software and associated documentation files (the "Software"), to deal    */
/* in the Software without restriction, including without limitation the rights     */
/* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell        */
/* copies of the Software, and to permit persons to whom the Software is            */
/* furnished to do so, subject to the following conditions:                         */
/* The above copyright notice and this permission notice shall be included in       */
/* all copies or substantial portions of the Software.                              */
/*                                                                                  */
/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR       */
/* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,         */
/* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE      */
/* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER           */
/* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,    */
/* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN        */
/* THE SOFTWARE.                                                                    */
/*                                                                                  */
/************************************************************************************/
/*                                                                                  */
/* AUTHOR  : Aby Louw                                                               */
/* DATE    : December 2009                                                          */
/*                                                                                  */
/************************************************************************************/
/*                                                                                  */
/* C convenience functions for saving/loading utterances in EBML format.            */
/*                                                                                  */
/*                                                                                  */
/*                                                                                  */
/************************************************************************************/

%define sutterance_ebml_save_DOCSTRING
"""
save_ebml(path)

Save the utterance to a file in the EBML format.

:param path: Full path and file name of the file where the utterance should be saved to.
:type path: string
:note: All the features in the utterance (including in the relations and items) must
       have appropriate EBML format serialization formatters registered.
"""
%enddef

%feature("autodoc", sutterance_ebml_save_DOCSTRING) sutterance_ebml_save;


%define sutterance_ebml_load_DOCSTRING
"""
load_ebml(path)

Load the utterance from a file in the EBML format.

:param path: Full path and file name of the file where the utterance should be loaded from.
:type path: string
:note: All the features in the utterance (including in the relations and items) must
       have appropriate EBML format serialization formatters registered.
"""
%enddef

%feature("autodoc", sutterance_ebml_load_DOCSTRING) sutterance_ebml_load;


/*
 * Do not delete these delimiters, required for SWIG
 */
%inline
%{
	void sutterance_ebml_save(SUtterance *utt, const char *path, s_erc *error)
	{
		SObjectSave(S_OBJECT(utt), path, "spct_utt", error);
	}


	SUtterance *sutterance_ebml_load(const char *path, s_erc *error)
	{
		SObject *loadedUtt;

		loadedUtt = SObjectLoad(path, "spct_utt", error);
		if (*error != S_SUCCESS)
			return NULL;

		return S_UTTERANCE(loadedUtt);
	}
/*
 * Do not delete this delimiter, required for SWIG
 */
%}


%pythoncode
%{
# add the functions to the Speect SUtterance class
setattr(speect.SUtterance, "load_ebml", staticmethod(sutterance_ebml_load))
setattr(speect.SUtterance, "save_ebml", sutterance_ebml_save)
%}
